### DISCUSSION TITLE
Mismatch in the number of expected tokens after applying tok2vec and PyTorchLSTM

### DISCUSSION THREAD

**Discussion by dkorenci** [explosion/spaCy#8538] [Help: Coding & Implementations] [answered]
Hi, I'm trying to implement a custom NER component and I came across an error in pipe.get_loss() function.
I create truth labels by simply iterating over the (reference) documents and then over the tokens within each
document, and concatenating tokens' one-hot vectors representing ground-truth NER labels.
In my understanding, this representation should be aligned with the computed scores for the same batch.
However, the scores, obtained by applying tok2vec -> PyTorchLSTM -> Softmax, are of different shape.
Shape of the truth is (91, 11), while the shape of the scores is (94, 11) - as if extra vectors are added 
in addition the the ones computed for the tokens. Can some of the components introduce these extra vectors?
The model building code is below (tok2vec is a standard spacy component - spacy.Tok2Vec.v2 + 
spacy.MultiHashEmbed.v1 + spacy.MaxoutWindowEncoder.v2)

```
bilstm = PyTorchLSTM(nO=lstm_width*2, nI=emb_width, bi=True, depth=num_layers) 
classification = Softmax(nO=num_labels, nI=lstm_width)
model = chain(tok2vec, with_padded(bilstm), with_array(classification))
```

Thank you for the help!

**Comment by svlandeg** [comment, marked as answer]
Could you provide a minimal code snippet that reproduces the error? That would help in trying to understand what exactly is going on...

**Reply by dkorenci** [reply]
Of course. I added the project to github, the code is here: https://github.com/dkorenci/ner_cro/blob/main/scripts/bilstm_tagger.py
The code breaks at line 74, in get_loss(), upon execution of "spacy project run train"

**Reply by dkorenci** [reply]
Some additional info, after some debugging. The thinc model does not change the shape/num.tokens, the problem is that
the same text is, it seems, tokenized in two different ways. One Doc object, with length of 54, is given as input to the thinc model, while the other, of length 51 (correct length) is passed to get_loss as part of an Example object, where the mismatch (with the scores - output of thinc model) happens. What confuses me is that the predict method is not called (and everything happens right at the start of training), so I wonder from where is the thinc model called, ie. where is the incorrectly tokenized Doc inserted.

**Reply by dkorenci** [reply]
Finally, I located the source of the error. One text is loaded from DocBin where the IOB corpus is stored, (created with nlp = spacy.blank("hr")). The other is, *I believe*, tokenized with the tokenizer in the pipeline, built from default-bilstm.cfg and configured with lang = "hr". The results differ. 
The first tokenization is:
[Predsjednik|ciparskih|Grka|Tassos|Papadopoulos|smatra|kako|je|planom|kojeg|je|prošle|godine|finalizirao|glavni|tajnik|UN|-|a|Kofi|Annan|strani|ciparskih|Turaka|i|Turskoj|dano|"|skoro|sve|što|su|htjeli|,|više|nego|i|m|je|bilo|potrebno|,|i|više|nego|je|pravedno|"|,|kazao|je|Prendergast|Vijeću|.]

The second tokenization (DocBin)
[Predsjednik|ciparskih|Grka|Tassos|Papadopoulos|smatra|kako|je|planom|kojeg|je|prošle|godine|finalizirao|glavni|tajnik|UN-a|Kofi|Annan|strani|ciparskih|Turaka|i|Turskoj|dano|"|skoro|sve|što|su|htjeli|,|više|nego|im|je|bilo|potrebno|,|i|više|nego|je|pravedno|"|,|kazao|je|Prendergast|Vijeću|.]

The difference is the tokenization of "UN-a" (the one that belongs to UN, ie United Nations)

Hope this clarifies things.

**Reply by adrianeboyd** [reply]
Ah, that makes sense as the source of the difference. An `Example` often has different tokenization in the reference doc than in the predicted doc, so you need to align the reference annotation with the predicted annotation in `get_loss`. `Example` has a number of built-in methods to help you do this and you can see examples in most of the internal `get_loss` functions: https://spacy.io/api/example#get_aligned

https://github.com/explosion/spaCy/blob/af9d98440714085ceee0c3ae0b92afcc63a87024/spacy/pipeline/tagger.pyx#L229-L236

`TrainablePipe.update` is where the predicted docs come into play. Does that make sense?

**Reply by dkorenci** [reply]
Great, using the alignment methods is the solution then. Thank you for the help @adrianeboyd @svlandeg

Just a note, I understand the flexibility that comes with using the alignments, what I find quite surprising 
is that for the same python environment and for the same language code ('hr'), I get two different tokenizations, 
one by running "spacy project run train" command, and other by manually creating Doc's and saving them to the DocBin.
Isn't it reasonable to expect these two tokenization to be the same?

**Reply by adrianeboyd** [reply]
You should get the same tokenization from `lang = "hr"` and `spacy.blank("hr")` as long as you're using the exact same version of spacy down to the patch release. Once your model is initialized, the exact version of spacy doesn't matter because the tokenizer settings are saved as part of the model and loaded from there instead of the library. As long as the spacy major.minor version is the same, the tokenization should be identical because we wouldn't modify the tokenizer algorithm between patch releases.

What's probably happening here is that your gold corpus provides its own tokenization that's not identical to `spacy.blank("hr")` and you're using `Doc(words=)`, which lets you provide any tokenization you want. Since training corpora are rarely going to have the exact same tokenization as the tokenizer, you'll need to deal with the alignment at some point.

https://github.com/dkorenci/ner_cro/blob/a5fc64c7d86ceb8fc98ff6b5378f48fcd5b184b1/scripts/prepare_hr500k.py#L109-L110

If you didn't provide `words` here, the tokenization would be identical, but a lot of the other annotation would be out-of-sync or you wouldn't even be able to create the doc. If you provide entities as character offsets you can omit `words`, but if you're providing word-level tags for some attribute, you need to provide `words`, too, when creating the `Doc` or the results are potentially misaligned.

**Reply by dkorenci** [reply]
You're right, that is exactly what is happening, this is a good example of an use-case for the alignment functionality. Thank you for the clarification and for the details of spacy.
